### PR TITLE
Fixes for macOS

### PR DESCRIPTION
**Pull request by ncough**
Partial fixes for getting builds to work on macOS. Many fixes are quite hacky, hopefully someone who knows what they are doing can fix them properly.

* Explicitly link abseil into `gtirb`, as a dependency of protobuf. Not clear why this isn't done automatically.
* Missing `#include <map>` in `gtirb-pprinter`.
* Disable concurrent souffle builds.
* Explicitly provide an implementation of `protoc` to `basil`, due to issues downloading and caching it.
* Fix a linking issue with `ocamlPackages.z3`.
* Adjust build paths for `python-gtirb`

### CODE REVIEW

**Comment by katrinafyi** [comment]
This is excellent! Thanks for your work. I hope Nix wasn't too frustrating :)

**Comment by ncough** [comment]
Nix has been mostly pain free. A bit confused on by `basil` build if you get a moment, it passes on my machine but fails on ci. Seems the ci doesn't run the `preConfigure` step.

**Comment by katrinafyi** [comment]
Basil build is fixed on my computer by https://github.com/katrinafyi/pac-nix/pull/18/commits/302dd59b05957c8904d2123a8ec9b9ddb2bcb945. This happened because the configure phase was not being run in the _dependency_ derivation, which mkSbtDerivation runs to download all dependencies (evidenced by the `basil-sbt-dependencies.tar.zst>` prefix in front of the log messages, instead of `basil>`). Within this derivation, configuring [is explicitly disabled](https://github.com/zaninime/sbt-derivation/blob/6762cf2c31de50efd9ff905cbcc87239995a4ef9/lib/dependencies.nix#L42).

Propagating the replacement command makes it work. I suspect some caching was at play to make it succeed locally.

**Comment by katrinafyi** [comment]
Re protobuf abseil linking errors: https://www.github.com/protocolbuffers/protobuf/issues/12637

**Comment by ncough** [comment]
From what I see locally, the only remaining issue is the determinism fix for ddisasm not working on macOS. If this isn't used, I suspect it would be far easier to disable the test.

**Comment by katrinafyi** [comment]
It looks really good now! Just one thing left, we should make the ddisasm test use a "nolibc" variant of the compiler so it doesn't spin cross-compiling libc. I can look into this if you want.

**Comment by ncough** [comment]
Thanks! I'll leave it to you. I was wondering if it would be possible to have Nix provide a full cross compiling environment on macOS anyway, as full tests aren't practical without it.

File: aslp/aslp-cpp.nix [hunk 1]
```diff
@@ -1,9 +1,13 @@
 { stdenv
 , cmake
 , asli
+, overrideCC
+, llvmPackages_17
 }:
 
-stdenv.mkDerivation {
+let buildStdenv = if stdenv.isDarwin then overrideCC stdenv llvmPackages_17.clang else stdenv; in
```
Comment (katrinafyi) [inline aslp/aslp-cpp.nix:8]: investigate better way. i assume this is just to obtain std::format

File: bap/bap.nix [hunk 2]
```diff
@@ -0,0 +1,11 @@
+{ stdenv, makeBinaryWrapper, symlinkJoin, orig-bap, z3 }:
+
+symlinkJoin {
+  inherit (orig-bap) version;
+  name = "bap-wrapped";
+  paths = [ orig-bap ];
+  nativeBuildInputs = [ makeBinaryWrapper ];
+  postBuild = ''
+    wrapProgram $out/bin/bap --prefix DYLD_FALLBACK_LIBRARY_PATH : ${z3.lib}/lib
```
Comment (katrinafyi) [inline bap/bap.nix:9]: simply move into bap package

File: basil/0001-basil-protoc-version.patch [hunk 3]
```diff
@@ -0,0 +1,13 @@
+diff --git a/build.sbt b/build.sbt
+index f1e6339f..2d9ede97 100644
+--- a/build.sbt
++++ b/build.sbt
+@@ -31,6 +31,8 @@ lazy val root = project
+ 
+ scalacOptions ++= Seq("-deprecation", "-feature")
+ 
++PB.protocExecutable := file("PROTOC_PLACEHOLDER")
```
Comment (katrinafyi) [inline basil/0001-basil-protoc-version.patch:9]: investigate possibility of upstreaming through a more flexible method

File: gtirb/ddisasm.nix [hunk 4]
```diff
@@ -28,6 +28,7 @@ stdenv.mkDerivation {
     rev = "9edfe9fe86910ef946de1db7a7ac41ce86bc31d0";
     hash = "sha256-xFW6J3jCCMtUqT25/zVfjjy5o3MoX1HXxIlRhjM6s8A=";
   };
+  patches = if stdenv.isDarwin then [ ./0001-ddisasm-disable-concurrent-souffle.patch ] else [];
```
Comment (katrinafyi) [inline gtirb/ddisasm.nix:31]: use null to avoid rebuild on x86

File: gtirb/ddisasm.nix [hunk 5]
```diff
@@ -52,17 +53,19 @@ stdenv.mkDerivation {
...
+  # Deterministic fix does not work on Darwin, just test to see if ddisasm even runs
+  passthru.tests.ddisasm-deterministic = runCommand
     "ddisasm-deterministic-test"
-    { nativeBuildInputs = [ ddisasm.deterministic jq ]; }
+    { nativeBuildInputs = [ ddisasm.deterministic jq clang-aarch64 ]; }
     ''
       mkdir -p $out && cd $out
       echo 'int main(void) { return 0; }' > a.c
-      $CC a.c
+      aarch64-unknown-linux-gnu-cc a.c
```
Comment (katrinafyi) [inline gtirb/ddisasm.nix:63]: investigate nolibc or pre-built binary

File: llvm-translator/alive2-regehr.nix [hunk 6]
```diff
@@ -15,5 +16,5 @@
   };
 
   patches = [ ];
-  CXXFLAGS = prev.CXXFLAGS or "" + " -Wno-error=maybe-uninitialized";
+  CXXFLAGS = prev.CXXFLAGS or "" + (if stdenv.isDarwin then "" else " -Wno-error=maybe-uninitialized");
```
Comment (katrinafyi) [inline llvm-translator/alive2-regehr.nix:19]: use lib.optional

File: overlay.nix [hunk 7]
```diff
@@ -13,8 +13,28 @@ let
 
       # llvm-translator packages 
       overlay_ocamlPackages = ofinal: oprev: {
-        ocaml-llvm-14 = ofinal.callPackage ./llvm-translator/ocaml-llvm.nix { libllvm = final.llvmPackages_14.libllvm; };
-        asl-translator = ofinal.callPackage ./llvm-translator/asl-translator.nix { llvm = ofinal.ocaml-llvm-14; };
+        # ctypes and ctypes-foreign v0.22.0 do not build on macOS
```
Comment (katrinafyi) [inline overlay.nix:16]: investigate...? if possible

File: .github/workflows/main.yml [hunk 8]
```diff
@@ -36,7 +37,7 @@ jobs:
           planter,
```
Comment (katrinafyi) [inline .github/workflows/main.yml:37]: remove planter from test list

**Review by katrinafyi — CHANGES_REQUESTED** [review]
todo notes for myself

**Comment by katrinafyi** [comment]
@ncough real quick about using Nix as a cross-compile  environment, this is definitely possible. but the main problem is the Nix-provided compilers are not hermetic - they will look for and may or may not find libraries on your system. It is especially dicey if the Nix compilers call a non-Nix tool which it finds through PATH.

The solutions I can conceptualise are:
-  run all compilations within a Nix sandbox - clunky, slow, difficult to integrate, adds Nix as a runtime dependency, and bloats Nix store on disk with compilation results
- use Nix compiler within a chroot / Docker - probably easiest. some trouble transporting files across the boundary 
- somehow manually isolate the Nix compiler, e.g. by clearing PATH in a wrapper script - hard

File: gtirb/ddisasm.nix (hunk 5 above)
Comment (katrinafyi) [reply gtirb/ddisasm.nix:63]: abandoned

**Review by katrinafyi — APPROVED** [review]

**Pull request merged**